design/src/rfcs/rfc0004_retry_behavior.md

```diff
@@ -0,0 +1,194 @@
+    - If `AWS_MAX_ATTEMPTS` is 1, retries will be disabled
+    - If `AWS_MAX_ATTEMPTS` is greater than 1, retries will be attempted at most as many times as is specified
+    - If the user creates the config with the `.disable_retries` builder method, retries will be disabled
+    - If the user creates the config with the `retry_config` builder method, retry behavior will be set according to the `RetryConfig` they passed
+1. The user creates a fluent client for the service they want to interact with and passes the config they created
+    - Provider precedence will determine what retry behavior is actually set, working like how `Region` is set
+1. The user calls an operation builder method on the client which constructs a request
+1. The user sends the request by awaiting the `send()` method
+1. The smithy client creates a new `Service` and attaches a copy of its retry policy
+1. The `Service` is `call`ed, sending out the request and retrying it according to the retry policy
+
+These changes will be made in such a way that they enable us to add the "adaptive" retry behavior at a later date without introducing a breaking change.
+
+Changes checklist
+-----------------
+
+- [ ] Create new Kotlin decorator `RetryConfigDecorator`
+  - Based on [RegionDecorator.kt](https://github.com/awslabs/smithy-rs/blob/main/aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/RegionDecorator.kt)
+- [ ] **Breaking changes:**
+  - [ ] Rename `smithy_client::retry::Config` to `StandardRetryConfig`
+  - [ ] Rename `smithy_client::retry::Config::with_max_retries` method to `with_max_attempts` in order to follow AWS convention
+- [ ] Create non-exhaustive `aws_types::retry_config::RetryConfig` enum wrapping structs that represent specific retry behaviors
+  - [ ] A `NoRetry` variant that disables retries. Doesn't wrap a struct since it doesn't need to contain any data
+  - [ ] A `Standard` variant that enables the standard retry behavior. Wraps a `StandardRetryConfig` struct.
+- [ ] Create `aws_config::meta::retry_config::RetryConfigProviderChain`
+- [ ] Create `aws_config::meta::retry_config::ProvideRetryConfig`
+- [ ] Create `EnvironmentVariableMaxAttemptsProvider` struct
+- [ ] Add `retry_config` method to `aws_config::ConfigLoader`
+- [ ] Update `AwsFluentClientDecorator` to correctly configure the max retry attempts of its inner `aws_hyper::Client` based on the passed-in `Config`
+- [ ] Add tests
+  - [ ] Test that setting retry_config to 1 disables retries
+  - [ ] Test that setting retry_config to `n` limits retries to `n` where `n` is a non-zero integer
+  - [ ] Test that correct precedence is respected when overriding retry behavior in a service-specific config
+  - [ ] Test that correct precedence is respected when overriding retry behavior in a shared config
+  - [ ] Test that setting invalid `max_attempts=0` with a `StandardRetryConfig` does not error and is ignored in favor of the default
```

Comment:
can you explain this a bit more? in what context would this be set? I'd expect this to error given the rest of the RFC

Reply:
As I understood it, if a higher-precendence config was set, that config would only override a lower precedence config value if the higher-precedence value was valid. Is that correct? I'm not entirely clear on how precedence and validity overlap 😳

Reply:
Higher-precedence values always override lower-precedence values even if they're invalid